oshmem/mca/spml/ucx/spml_ucx.h

```diff
@@ -101,33 +101,48 @@ typedef struct mca_spml_ucx_ctx_array {
     mca_spml_ucx_ctx_t       **ctxs;
 } mca_spml_ucx_ctx_array_t;
 
+typedef struct mca_spml_ucx_rkey {
+    ucp_rkey_h rkey;
+    uint16_t refcnt;
+} mca_spml_ucx_rkey_t;
+
+typedef struct mca_spml_ucx_rkey_store {
+    int                 size;
+    int                 count;
+    mca_spml_ucx_rkey_t *array;
+    ucp_worker_h        worker;
+} mca_spml_ucx_rkey_store_t;
+
 struct mca_spml_ucx {
-    mca_spml_base_module_t   super;
-    ucp_context_h            ucp_context;
-    int                      num_disconnect;
-    int                      heap_reg_nb;
-    bool                     enabled;
+    mca_spml_base_module_t    super;
+    ucp_context_h             ucp_context;
+    int                       num_disconnect;
+    int                       heap_reg_nb;
+    bool                      enabled;
     mca_spml_ucx_get_mkey_slow_fn_t get_mkey_slow;
-    char                     ***remote_addrs_tbl;
-    mca_spml_ucx_ctx_array_t active_array;
-    mca_spml_ucx_ctx_array_t idle_array;
-    int                      priority; /* component priority */
-    shmem_internal_mutex_t   internal_mutex;
-    pthread_mutex_t          ctx_create_mutex;
+    char                      ***remote_addrs_tbl;
+    mca_spml_ucx_ctx_array_t  active_array;
+    mca_spml_ucx_ctx_array_t  idle_array;
+    int                       priority; /* component priority */
+    shmem_internal_mutex_t    internal_mutex;
+    pthread_mutex_t           ctx_create_mutex;
     /* Fields controlling aux context for put_all_nb SPML routine */
-    bool                     async_progress;
-    int                      async_tick;
-    opal_event_base_t        *async_event_base;
-    opal_event_t             *tick_event;
-    mca_spml_ucx_ctx_t       *aux_ctx;
-    pthread_spinlock_t       async_lock;
-    int                      aux_refcnt;
-    unsigned long            nb_progress_thresh_global;
-    unsigned long            nb_put_progress_thresh;
-    unsigned long            nb_get_progress_thresh;
-    unsigned long            nb_ucp_worker_progress;
-    unsigned int             ucp_workers;
-    unsigned int             ucp_worker_cnt;
+    bool                      async_progress;
+    int                       async_tick;
+    opal_event_base_t         *async_event_base;
+    opal_event_t              *tick_event;
+    mca_spml_ucx_ctx_t        *aux_ctx;
+    pthread_spinlock_t        async_lock;
+    int                       aux_refcnt;
+    unsigned long             nb_progress_thresh_global;
+    unsigned long             nb_put_progress_thresh;
+    unsigned long             nb_get_progress_thresh;
+    unsigned long             nb_ucp_worker_progress;
+    unsigned int              ucp_workers;
+    unsigned int              ucp_worker_cnt;
+    mca_spml_ucx_rkey_store_t rkey_store;
+    bool                      symmetric_rkey;
+    int                       symmetric_rkey_size;
```

Comment:
do we need to maintain it per `mca_spml_ucx_ctx_t` (i.e. per worker)?

Reply:
I am assuming we only care about first context (in practice only one `shmem_ctx_create()` call), and we do not want to create symmetric store for `aux_ctx` or other instances.

Reply:
actually, we want to create symmetric store for all contexts

Reply:
fixed